jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt

```diff
@@ -0,0 +1,324 @@
+ *
+ * 'JetBrains.Rider.Debugger.Launcher' app is hosted on github repo - https://github.com/JetBrains/JetBrains.Rider.Debugger.Launcher.
+ * The app is packed to NuGet package and is accessible in Rider SDK in '<SDKPath>/lib/ReSharperHost/' directory.
+ */
+class DotNetSamDebugSupport : SamDebugSupport {
+    companion object {
+        private val logger = getLogger<DotNetSamDebugSupport>()
+
+        const val DEBUGGER_LAUNCHER_NAME = "JetBrains.Rider.Debugger.Launcher"
+        private const val DEBUGGER_MODE = "server"
+
+        private const val REMOTE_DEBUGGER_DIR = "/tmp/lambci_debug_files"
+        private const val REMOTE_NETCORE_CLI_PATH = "/var/lang/bin/dotnet"
+        private const val REMOTE_LAMBDA_COMPILED_PATH = "/var/runtime/MockBootstraps.dll"
+    }
+
+    private val debuggerAssemblyFile =
+        RiderEnvironment.getBundledFile(DebuggerWorkerPlatform.AnyCpu.assemblyName)
+
+    private val debuggerBinDirectory = debuggerAssemblyFile.parentFile
+
+    /**
+     * Check whether the JatBrains.Rider.Worker.Launcher app (that is required to run Debugger) is downloaded into Rider SDK.
+     */
+    override fun isSupported(): Boolean {
+        val debugLauncherFile = File(debuggerBinDirectory, "$DEBUGGER_LAUNCHER_NAME.dll")
+
+        val debuggerLauncherExists = debugLauncherFile.exists()
+        if (!debuggerLauncherExists) {
+            logger.error { "$DEBUGGER_LAUNCHER_NAME runnable does not exists" }
+        }
+        return debuggerLauncherExists
+    }
+
+    override fun patchCommandLine(debugPort: Int, state: SamRunningState, commandLine: GeneralCommandLine) {
+        val debugArgs = StringBuilder()
+            .append("$REMOTE_DEBUGGER_DIR/$DEBUGGER_LAUNCHER_NAME.dll ")
+            .append("$REMOTE_DEBUGGER_DIR/runtime.sh ")
+            .append("$REMOTE_DEBUGGER_DIR/${debuggerAssemblyFile.name} ")
+            .append("$DEBUGGER_MODE ")
+            .append("$debugPort ")
+            .append("$debugPort")
+            .toString()
+
+        commandLine.withParameters("--debugger-path")
+                .withParameters(debuggerBinDirectory.path)
+                .withParameters("--debug-args")
+                .withParameters(debugArgs)
+
+        super.patchCommandLine(debugPort, state, commandLine)
+    }
+
+    override fun createDebugProcess(
+        environment: ExecutionEnvironment,
+        state: SamRunningState,
+        debugPort: Int
+    ): XDebugProcessStarter? {
+        throw UnsupportedOperationException("Use 'createDebugProcessAsync' instead")
+    }
+
+    override fun createDebugProcessAsync(
+        environment: ExecutionEnvironment,
+        state: SamRunningState,
+        debugPort: Int
+    ): Promise<XDebugProcessStarter?> {
+
+        val promise = AsyncPromise<XDebugProcessStarter?>()
+
+        try {
+            val exitValue = Runtime.getRuntime().exec("docker ps").waitFor()
```

Comment:
We should probably put this in the base since we always need docker. I like the check.

Reply:
Moved to SamDebugger#run.